### PR TITLE
Orthotropic material preparation

### PR DESCRIPTION
**Pull request by BenceVirtonomy**

### CODE REVIEW

File: SPHINXsys/src/shared/materials/elastic_solid.cpp [hunk 1]
```diff
@@ -129,6 +129,54 @@ namespace SPH {
...
+	void OrthotropicSolid::CalculateA0()
+	{
+		A_[0] = SimTK::outer(a_[0], a_[0]);
+		A_[1] = SimTK::outer(a_[1], a_[1]);
+		A_[2] = SimTK::outer(a_[2], a_[2]);
+		
+	}
+	void OrthotropicSolid::CalculateAllMu()
+	{
+		G_[0]=2/(Mu_[0]+Mu_[1]);
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.cpp]: G is input and we need to get all the Mu

File: SPHINXsys/src/shared/materials/elastic_solid.h [hunk 2]
```diff
@@ -172,6 +172,49 @@ namespace SPH {
...
+	* @brief Ortothropic solid
+	*/
+	class OrthotropicSolid : public LinearElasticSolid
+	{
+	public:
+		OrthotropicSolid() : LinearElasticSolid() 
+		{
+			material_name_ = "OrthotropicSolid";
+		};
+		OrthotropicSolid(Real rho_0, std::array<Vecd, 3> a, std::array<Real, 3> E, std::array<Real, 3> G,std::array<Real, 3> poisson)
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.h]: we have to call all these functions in the constructor here:
CalculateAllMu();
CalculateAllLambda();
CalculateA0();

File: SPHINXsys/src/shared/materials/elastic_solid.cpp [hunk 3]
```diff
@@ -129,6 +129,54 @@ namespace SPH {
...
+		G_[1]=2/(Mu_[1]+Mu_[2]);
+		G_[2]=2/(Mu_[2]+Mu_[0]);
+	}
+	void OrthotropicSolid::CalculateAllLambda()
+	{
+		Matd Complience= Matd(Vecd(1/E_[1], -poisson_[1]/E_[1], -poisson_[2]/E_[1]),
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.cpp]: please put comments and explain the indeces: possions[1] means nu_12 etc.

File: SPHINXsys/src/shared/materials/elastic_solid.cpp [hunk 4]
```diff
@@ -129,6 +129,54 @@ namespace SPH {
...
+					Vecd(-poisson_[1]/E_[2], 1/E_[2], -poisson_[3]/E_[2]),
+					Vecd(-poisson_[2]/E_[3], -poisson_[3]/E_[3], 1/E_[3]));
+
+		// Matd M= Matd(Vecd (Lambda_[0]+2*Mu_[0], Lambda_[3], Lambda_[4]),
+		// 		Vecd(Lambda_[3], Lambda_[1]+2*Mu_[1], Lambda_[5]),
+		// 		Vecd(Lambda_[4], Lambda_[5], Lambda_[2]+2*Mu_[2]));
+
+		Matd Compliance_inv= SimTK::inverse(Complience);
+		
+		Lambda_[0]=Compliance_inv[0][0]-2*Mu_[0];
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.cpp]: same here, explain which Lambda is which

File: SPHINXsys/src/shared/materials/elastic_solid.cpp [hunk 5]
```diff
@@ -129,6 +129,54 @@ namespace SPH {
...
+		Lambda_[1]=Compliance_inv[1][1]-2*Mu_[1];
+		Lambda_[2]=Compliance_inv[1][1]-2*Mu_[2];
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.cpp]: should be Compliance_inv[2][2]

File: SPHINXsys/src/shared/materials/elastic_solid.h (hunk 2 above)
Comment (BenceVirtonomy) [reply SPHINXsys/src/shared/materials/elastic_solid.h]: and we should also move the constructor to the .cpp and only declare it here

File: SPHINXsys/src/shared/materials/elastic_solid.h [hunk 6]
```diff
@@ -172,6 +172,49 @@ namespace SPH {
...
+		
+		
+	
+		/** second Piola-Kirchhoff stress related with green-lagrangian deformation tensor */
+		virtual Matd ConstitutiveRelation(Matd& deformation, size_t particle_index_i) override;
+		/** Volumetric Kirchhoff stress determinate */
+		virtual Real VolumetricKirchhoff(Real J) override;
+
+		protected:
+		//Vecd f0_, s0_; 				/**< Reference fiber and sheet directions as basic parameter. */
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.h]: delete these, that we don't need

File: SPHINXsys/src/shared/materials/elastic_solid.h [hunk 7]
```diff
@@ -172,6 +172,49 @@ namespace SPH {
...
+		//Matd f0f0_, s0s0_, f0s0_;	/**< Tensor products of fiber and sheet directions as basic parameter.. */
+		std::array<Vecd, 3> a_;
+		std::array<Real, 3> E_;
+		std::array<Real, 3> G_;
+		std::array<Real, 3> poisson_;
+		Matd A_[3];
+		Real Lambda_[6]; //123-main 456-side
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.h]: explain the indices a bit more detailed

File: SPHINXsys/src/shared/materials/elastic_solid.h [hunk 8]
```diff
@@ -182,12 +182,13 @@ namespace SPH {
 	public:
 		OrthotropicSolid() : LinearElasticSolid()
```
Comment (EttoTor) [inline SPHINXsys/src/shared/materials/elastic_solid.h]: https://help.febio.org/FEBio/FEBio_tm_2_7/FEBio_tm_2-7-Subsection-5.2.2.html#prev

File: SPHINXsys/src/shared/common/small_vectors.cpp [hunk 9]
```diff
@@ -214,6 +214,31 @@ namespace SPH {
 		return diag;
 	}
 	//=================================================================================================//
+	Real CalculateDoubleDotProduct(Mat2d Matrix1, Mat2d Matrix2 )
```
Comment (chuvirtonomy) [inline SPHINXsys/src/shared/common/small_vectors.cpp:217]: this function looks problematic for me. how do you know the size of the matrix is (2,2)? can not be larger than 2x2?

File: SPHINXsys/src/shared/common/small_vectors.cpp [hunk 10]
```diff
@@ -214,6 +214,31 @@ namespace SPH {
...
+		for(int i=0; i < 2; i++)
+		{
+			for(int j=0; j < 2; j++)
+			{
+				product += Matrix1[i][j] * Matrix2[i][j];
+			}
+		}
+		return product;
+	}
+	Real CalculateDoubleDotProduct(Mat3d Matrix1, Mat3d Matrix2 )
```
Comment (chuvirtonomy) [inline SPHINXsys/src/shared/common/small_vectors.cpp:229]: This is also strange for me. It does not seems to be you are doing dot product.

**Review by chuvirtonomy — CHANGES_REQUESTED** [review]

**Review by chuvirtonomy — APPROVED** [review]

**Pull request merged**